middle_end/flambda2/docs/backend_exceptions.md

```diff
@@ -0,0 +1,241 @@
+a trap frame, after the body (if the body doesn't raise) the execution resumes
+with the base stack, and if the body raises execution goes to the handler, with
+the base stack (and when the handler finishes execution flows to the same code
+as after the body, with the base stack).
+
+## Problematic cases
+
+This all works well as long as all code paths go through the `Lpushtrap` and
+`Lpoptrap` instructions in a well-parenthesized manner. However, this invariant
+can be broken by the static exceptions used in the compiler representations
+from Lambda and onwards. In particular, although it only occurs in specific
+circumstences, it is allowed to have inside the body of a `try...with`
+structure a sub-expression that raises a static exception whose handler is
+outside the body of the `try...with`. This looks like a simple detail, as
+the code that emits the traps simply needs to insert traps around these
+static raising constructs, but this is actually also important for earlier
+passes as some optimisations involving static exceptions that would be correct
+if both the handler and the raise are in the same exception context become
+incorrect if the contexts differ.
+
+To deal with this, the relevant optimisations in the compiler carry a `try_depth`
+argument that allows to check if the optimisation is allowed or not. This of
+course means that someone writing a new optimisation without thinking about this
+slightly uncommon pattern is bound to encounter the same problems again.
+
+In flambda2, we chose to deal with this by computing during translation from
+lambda what kind of (abstract) trap actions are associated with each static
+raise (or `Apply_cont` in flambda2 terminology), so all transformation passes
+on the flambda2 representation are forced to deal with the potential trap
+context updates.
+
+## Trap context and CFG representations
+
+The main property needed for correct translation to machine code is that each
+piece of code knows exactly what trap context it will be evaluated in.
+This is to guarantee that access to stack-based pseudo-registers can be
+compiled to a statically known offset from the stack pointer, even through
+pieces of codes that modify the stack pointer (the trap instructions).
+
+For the current implementation, this is fairly straightforward: all intermediate
+representations except the very last ones (linear and the generated assembly)
+have some kind of expression structure where the only way to change the trap
+context is to enter or exit a trywith block.
+
+But there have been some experiments (in particular the flambda2 project, but
+not only) that use a representation of the code that is either explicitly
+or implicitly a control-flow graph. For those representations, the ability
+to re-order the nodes (or instruction blocks) of the graph is an important
+property, and at least for flambda2 it's very easy to end up with a graph
+for which there is no straightforward transformation back into a structured
```

Comment:
Out of curiosity, how important is this property?

Reply:
This was written to justify why we couldn't easily reuse the existing `Ctrywith`/`Itrywith` instruction in the backend. If we could rebuild structured expressions, we wouldn't have needed to introduce explicit traps so early in the backend.

Reply:
I see, thanks; what I was actually wondering was along the lines
of "Once we have a CFG, do we need to go back to a _structured_
representation? it looks like at this point we only want to emit
code."

Reply:
I wasn't aware of the work on CFG representations for Linear instructions when I wrote this document. Of course we don't need to go back to a structured representation at that point, but the Flambda2 terms (which are also similar to control-flow graphs) need to go through the normal compilation pipeline (Cmm, Mach, and so on) which uses structured expressions.

Reply:
Understood, thanks for the contextualisation.